사용자 회원 등록 및 수정 API 기능 및 테스트 코드 작성

shinsanghooon opened the pull request:
다음 이슈에 대한 PR입니다. 
- #8
- #9

shinsanghooon committed:
<chore>: data.sql 파일 추가

위 변경 사항이 main 브랜치에 아직 merge가되지 않아서 작업을 계속하기
위해 임시로 파일을 동일하게 변경

shinsanghooon committed:
feat [#8]: 회원 가입 API 구현

1. 사용자 정보를 입력받아 회원가입 API 구현

Related to #8

shinsanghooon committed:
feat [#8]: 비밀번호 암호화 기능 구현

1. 비밀번호를 단방향으로 암호화하여 DB에 저장
2. UserRequestDto 생성자를 통해 인스턴스가 생성될 때 암호화된 비밀번호
   값을 세팅

Related to #8

shinsanghooon committed:
feat [#8]: 패스워드 재확인 기능 구현

1. 입력한 패스워드를 다시 입력하게 하여 패스워드 일치 여부를 확인
2. 입력된 패스워드가 일치하지 않는 경우 예외를 던짐

shinsanghooon committed:
feat [#8]: 입력값에 대한 Validation 구현

1. 스프링 Validation 의존성 추가
2. 회원가입 입력값에 대한 Validation 처리

shinsanghooon committed:
feat [#8]: javadoc 작성

1. 코드를 읽는 사람이 빠르게 파악하는데 도움이 되도록 설명 추가하였음

shinsanghooon committed:
feat [#8]: 회원 가입시 발생할 수 있는 예외 처리

1. 이름, 이메일, 전화번호에 입력값에 대한 포맷 관련 예외 처리 진행
2. 이메일 중복여부 확인 API 개발
3. 패스워드와 패스워드 확인에 입력된 값 일치 여부 확인 기능 추가
4. 던져진 예외는 `@ControllerAdvice`와 `@ExceptionHandler` 적용하여
   처리하도록 구현함

shinsanghooon committed:
feat [#8]: 응답값을 User 객체를 UserResponseDto 객체로 변경

1. User 객체를 응답값에 그대로 노출시키는 것보다 UserResponseDto 객체를 사용해 필요한 값만 응답할 수 있게 변경하였음

shinsanghooon committed:
feat [#8]: 응답값을 User 객체를 UserResponseDto 객체로 변경

1. User 객체를 응답값에 그대로 노출시키는 것보다 UserResponseDto 객체를
   사용해 필요한 값만 응답할 수 있게 변경하였음

shinsanghooon committed:
chore [#8]: 테스트 라이브러리 추가

1. assertj를 사용을 위한 라이브러리 추가
2. parameterizedTest를 사용을 위한 라이브러리 추가

shinsanghooon committed:
feat [#8]: UserService 단위 테스트 작성

1. UserService 단위 테스트 작성

shinsanghooon committed:
chore [#8]: 사용하지 않는 패키지 정리

1. 사용하지 않는 패키지 정리

shinsanghooon committed:
feat [#9]: 회원 정보 수정 API

1. 이메일은 아이디로 사용하기 때문에 변경할 수 없음
2. 이름, 패스워드, 전화번호만 변경이 가능

shinsanghooon committed:
feat [#9]: 회원정보 수정 테스트 코드 추가

1. 회원정보 수정 테스트 코드 추가

shinsanghooon committed:
fix: git conflict 해결

Merge branch 'main' into feature/user

kyucumber commented on build.gradle:
```diff
@@ -15,6 +15,10 @@ repositories {
     implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:2.2.2'
+
+    // 스프링 Validation 추가
+    implementation("org.springframework.boot:spring-boot-starter-validation")
```
안그래도 요거 말씀드리려고 했는데 이미 적용하셨네요 👍

kyucumber commented on src/main/java/com/shyb/evcharging/user/api/UserController.java:
```diff
@@ -0,0 +1,81 @@
+ */
+
+@RestController
+@RequestMapping("/users")
```
API가 변하는 부분(버저닝)과 외부 API, 내부 API 구분을 위해 아래와 같이 사용하는 경우도 있습니다. 참고 부탁드려요.

```suggestion
@RequestMapping("/api/v1/users")
```

내부는 `webapi`, 외부는 `api` 요런식으로 분리해서 사용할수도 있어요.
그리고 인증이 적용될 경로와 적용되지 않을 경로(헬스체크 등)을 구분할 필요가 있는 경우가 있어서 앞에 위와 같이 접두사가 붙는게 좋을 것 같아요.

kyucumber commented on src/main/java/com/shyb/evcharging/user/api/UserController.java:
```diff
@@ -0,0 +1,81 @@
+
+    private final UserService userService;
+
+    /**
+     * 사용자를 등록한 후 리턴합니다.
+     *
+     * @param user 생성 요청 사용자 정보
+     * @return 등록된 사용자
+     */
```
개인적으로는 java doc 관련 주석을 모든 메소드에 남길 필요는 없다고 생각합니다.

REST API를 잘 설계했다면 HTTP 메소드나 자원 명, 메소드 명과 반환 타입을 통해 어떤 자원에 대해 다루는지 충분히 이해할 수 있다고 생각해요.

shinsanghooon replied:
컨트롤러에서는 말씀해주신 것처럼 REST API를 잘 설계하면 충분히 이해가 가능한 부분이라고 생각됩니다. 적으면서도 너무 뻔한걸 적는다는 느낌도 받긴 했지만 그래도 작성해두었습니다...! ㅎㅎㅎ

kyucumber commented on src/main/java/com/shyb/evcharging/user/api/UserController.java:
```diff
@@ -0,0 +1,81 @@
+     * @param email 이메일 중복 요청 DTO
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @PostMapping("/email/check")
```
중복 여부라는 자원의 조회로 본다면 `@GetMapping`으로도 사용할 수 있지 않을까요?

단순 이메일의 중복 여부를 체크하는 기능이라면 보안적인 문제도 없을 것 같아 `@PostMapping`을 사용하지 않아도 괜찮을 것 같아요.

shinsanghooon replied:
조회의 관점에서 바라보니 `@GetMapping` 이 의미적으로 조금 더 알맞는 거 같습니다.

kyucumber commented on src/main/java/com/shyb/evcharging/user/api/UserController.java:
```diff
@@ -0,0 +1,81 @@
+     * @return 조회된 사용자
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping("/users/{id}")
```
`/users/users/{id}`로 users가 중복해서 URL에 표시되는 것 같아요.

kyucumber commented on src/main/java/com/shyb/evcharging/user/application/UserService.java:
```diff
@@ -0,0 +1,94 @@
+        String email = emailDuplicateCheckRequestDto.getEmail();
+        Optional<UserResponseDto> user = userRepository.findByEmail(email);
+        if (user.isPresent()) {
+            throw new EmailDuplicateException();
```
모든 예외를 커스텀으로 만들 필요는 없다고 생각해요.
이미 존재하는 표준 예외에 적절한 메시지를 통해 어떤 예외 상황인지를 나타내주어도 충분하지 않을까요?

아래 내용을 한번 참고해보세요.

https://insight-bgh.tistory.com/470

kyucumber commented on src/main/java/com/shyb/evcharging/user/application/UserService.java:
```diff
@@ -0,0 +1,94 @@
+        }
+        user.passwordEncrypt();
+
+        return userRepository.save(user.toEntity());
```
DTO의 불변이 해당 메소드 호출로 인해 깨졌고, DTO와 PasswordEncoder간의 결합이 생겼네요.

크게 중요한 부분은 아니지만 아래와 같이 정의하면 불변을 유지할 수 있지 않을까요?

```suggestion
        String encryptedPassword = ~~
        return userRepository.save(user.toEntity(encryptedPassword));
```

kyucumber commented on src/main/java/com/shyb/evcharging/user/application/UserService.java:
```diff
@@ -0,0 +1,94 @@
+     * @param id 사용자 식별자
+     * @return 조회된 사용자
+     */
+    public UserResponseDto findById(Long id) {
+        return userRepository.findById(id).orElseThrow(UserNotFoundException::new);
+    }
```
findById 메소드 내부에서 해당 아이디로 조회하지 못하는 경우 예외를 던지고 있네요.

JPA의 repository의 findById는 어떻게 동작하는지 찾아보시고 get과 find 접두사가 붙은 메소드들에 대해 어떤 의미 차이가 있을지 한번 고민해보세요

shinsanghooon replied:
의미상으로 한 번 생각해본 바로는  get은 원하는 값을 얻기 위한 행위이고, find는 값을 얻는 것보다 값이 있는지 확인하는게 더 주요한 목적인 것 같다는 생각이 들었습니다. 

그리고 살짝 찾아보았습니다. JPA에서 관련 메서드들은 아래와 같은 특징을 가집니다.
- `findById` 의 경우, 리턴타입으로 `Optional<T>`를 가지며 값이 없는 경우 `Optional.empty()`를 리턴하게 됩니다. 즉 값이 없어도 예외가 던져지지 않습니다. (예외는 id값에 null을 넣거나 하는 경우 발생) 
- `getOne` 의 경우, 리턴타입은 그냥 T 입니다. 때문에 값이 없을 때는 예외가 던져지게 됩니다. 
- 추가로 getOne 메소드의 경우, 내부적으로 lazy loading 를 지원하고 findById의 경우에는 바로 실제 엔티티 객체를 조회한다는 차이점이 존재합니다. 

그 외에도 데이터를 가져오는 시간의 정도, 로직의 복잡도 같은 이유에 따라서도 get과 find를 구분해서 사용하기는 사례도 있는 것 같았습니다. 

만약 jpa 문서 내용이 find와 get을 구분해서 사용하는 관례라면 제가 작성한 코드는 find 메소드는 예외를 던지고 있으니 관례를 어겼다고 볼 수도 있겠네요..? 허허ㅎㅎ

kyucumber commented on src/main/java/com/shyb/evcharging/user/application/UserService.java:
```diff
@@ -0,0 +1,94 @@
+     * @param emailDuplicateCheckRequestDto 가입 요청된 이메일 DTO
+     * @return 가입 가능한 이메일 주소인 경우 true를 리턴합니다.
+     */
+    public boolean checkEmailDuplicate(EmailDuplicateCheckRequestDto emailDuplicateCheckRequestDto) {
```
check의 경우 stdlib에서도 보통 void 형태의 메소드로 정의되었던 걸로 기억합니다.

실제로 Kotlin의 check라는 stdlib 함수도 체크 후 예외를 던지지 boolean 값을 리턴하는 형태는 아니에요.

대부분 check~~와 같이 네이밍되었다면 조건을 체크 후 예외를 던지는 형태로 생각할 것 같은데요 boolean 타입의 반환을 예상할 수 있는 메소드명으로 변경하면 어떨까요?

shinsanghooon replied:
check ~~ 로 메소드 명은 유지하고 대신 예외를 던지는 형태로 수정하였습니다.

kyucumber commented on src/main/java/com/shyb/evcharging/user/domain/User.java:
```diff
@@ -0,0 +1,48 @@
+@Getter
+@Setter
+@NoArgsConstructor
+public class User{
```
코드 포맷팅이 제대로 되어있지 않네요.

이후에 시간이 나시면 프리커밋 훅을 통해 lint를 자동으로 돌려 포맷팅을 맞춰주거나 체크해주는 부분을 넣어보아도 좋을 것 같습니다.

kyucumber commented on src/main/java/com/shyb/evcharging/user/dto/ErrorResponse.java:
```diff
@@ -0,0 +1,21 @@
+ */
+@Getter
+@Setter
+public class ErrorResponse {
```
Dto의 네이밍 컨벤션은 `~~Dto`로 끝나는 거 같은데 이 클래스는 컨벤션이 맞지 않네요.

중요한 부분은 아니지만 클래스의 일관성을 해치기 때문에 다른 부분의 `~~Dto`를 제거하거나 요 클래스에 Dto를 추가해주시면 좋을 것 같아요.

kyucumber commented on src/main/java/com/shyb/evcharging/user/dto/UserModifyRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+@Getter
+@Setter
+@NoArgsConstructor
+public class UserModifyRequestDto{
+
+    @Size(max=20, message="이름의 최대 길이는 20자 입니다.")
```
코드 포맷팅이 제대로 안 돈 것 같아요. IDE 설정을 한번 확인 부탁드립니다.

설정이 정상적으로 되어 있다면 코드 포맷을 자주 맞춰주시면 좋을 것 같아요.

별 거 아니지만 협업을 할 때 이 모든게 file change로 잡히게 되고 코드리뷰때 불필요한 file change로 인해 중요한 부분을 놓칠 수도 있습니다.

kyucumber commented on src/main/java/com/shyb/evcharging/user/exception/UserExceptionHandler.java:
```diff
@@ -0,0 +1,71 @@
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    @ResponseStatus(code= HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleMethodArgumentNotValid(MethodArgumentNotValidException ex) {
+        List<String> errors = ex.getBindingResult()
+            .getFieldErrors()
+            .stream()
+            .map(DefaultMessageSourceResolvable::getDefaultMessage)
+            .collect(Collectors.toList());
+
+        return new ErrorResponse(errors.get(0));
```
현재 0번째만 가져오고 있어서 여러 필드에 동시에 오류가 난 경우는 값을 식별하기가 어려울 것 같아요.

ErrorResponse의 메시지를 리스트로 변경하거나 Collectors.joining()인가 그런게 있었던 것 같은데 에러 메시지를 이어 붙여 반환하면 어떨까요?

kyucumber commented on src/main/java/com/shyb/evcharging/user/repository/UserSearchCond.java:
```diff
@@ -0,0 +1,27 @@
+
+@Getter
+@Setter
+public class UserSearchCond {
```
Cond라는 네이밍은 어떤 의미인가요?

shinsanghooon replied:
Condition을 의미하며 사용자 조회를 할 때 조건을 담고 있는 객체이고 UserMapper.xml에서 사용됩니다. 처음에 MyBatis 세팅할 때 예시로 사용했던건데 지금은 필요하지 않아서 다음 푸시에 정리해서 올리도록 하겠습니다.

kyucumber replied:
요건 좀 개인적인 의견일 수 있는데 정말로 너무 긴게 아니면 Cond같은 축약어보다는 Condition을 사용하는게 좋은 것 같아요.

보는 사람에 따라 정말 상상도 못한 의미로 해석해서 오해나 버그가 생기는 경우도 있었거든요.

kyucumber commented on src/main/java/com/shyb/evcharging/user/repository/mybatis/UserMapper.java:
```diff
@@ -0,0 +1,25 @@
+
+    void save(User user);
+
+    List<UserResponseDto> findAll(UserSearchCond userSearch);
+
+    Optional<UserResponseDto> findById(Long id);
+
+    Optional<UserResponseDto> findByEmail(String email);
+
+    Long update(Long id, @Param("userModifyRequestDto") UserModifyRequestDto userModifyRequestDto);
```
개인적으로는 UserMapper를 포함해서 Repository 레이어까지 Dto가 침범하는건 좋지 않은 구현이라고 생각해요.

JPA로 예를 들면 Repository 레이어에서는 User Entity를 반환하겠죠.

물론 변환하거나 하는 비용이 상당하긴 하겠지만 컨트롤러 레이어에서 데이터를 주고받을 목적으로 만든 Dto가 하위 레이어까지 침범하면 Dto의 변경에 모든 레이어가 영향을 받을 것 같습니다.

변환하는 비용이 너무 크기 때문에 예시로 보여드릴때도 그렇게 짜지는 않았지만 정말 빡빡하게 규칙을 잡는다면 Service가 Dto를 반환하는것도 문제가 될 수 있긴 해요.

kyucumber commented on src/main/java/com/shyb/evcharging/utils/PasswordEncoder.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * 패스워드 암호화 작업을 처리합니다.
+ */
+public class PasswordEncoder {
```
직접 구현하셨네요.

이후에 시큐리티를 적용하거나 사용할 일이 있다면 아래 클래스를 사용하셔도 좋을 것 같습니다.

https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.html

kyucumber commented on src/main/resources/com/shyb/evcharging/user/repository/mybatis/UserMapper.xml:
```diff
@@ -0,0 +1,46 @@
+        values (#{name}, #{password}, #{phone}, #{email})
+    </insert>
+
+    <select id="findById" resultType="UserResponseDto">
```
위에 적었지만 가급적 Repository 레이어에서 Dto를 반환하는걸 지양해주셨으면 해요.

kyucumber commented on src/test/java/com/shyb/evcharging/user/application/UserServiceTest.java:
```diff
@@ -0,0 +1,211 @@
+    String VALID_NAME = "홍길동";
+
+    String VALID_PHONE = "010-1234-5678";
+
+    String VALID_EMAIL = "abcd@gmail.com";
+    String EXISTING_EMAIL = "existing@gmail.com";
+    String NOT_EXISTING_EMAIL = "non_exsting@gmail.com";
+
+    String PASSWORD = "12345";
+    String CONFIRM_PASSWORD = "12345";
+    String CONFIRM_PASSWORD_WITH_MISMATCH = "55555";
+
+    UserResponseDto userResponseDto;
+    UserRequestDto userRequestDTO;
+    UserRequestDto userRequestDTOWithPasswordMismatch;
+
+    @BeforeEach
+    void setUp() {
+        userResponseDto = UserResponseDto.builder()
+            .id(USER_ID)
+            .name(VALID_NAME)
+            .phone(VALID_PHONE)
+            .email(VALID_EMAIL)
+            .build();
+
+        userRequestDTO = UserRequestDto.builder()
+            .name(VALID_NAME)
+            .phone(VALID_PHONE)
+            .email(VALID_EMAIL)
+            .password(PASSWORD)
+            .confirmPassword(CONFIRM_PASSWORD)
+            .build();
+
+        userRequestDTOWithPasswordMismatch = UserRequestDto.builder()
+            .name(VALID_NAME)
+            .phone(VALID_PHONE)
+            .email(VALID_EMAIL)
+            .password(PASSWORD)
+            .confirmPassword(CONFIRM_PASSWORD_WITH_MISMATCH)
+            .build();
```
테스트를 위해 준비해야하는 데이터들이 많고 이런 코드가 많아짐에 따라 테스트 클래스의 상당부분을 차지해 가독성이 떨어집니다.

TestFixture에 대한 내용을 검색하고 테스트를 위한 데이터를 만들기 위한 별도의 클래스를 만들어보셔도 좋을 것 같아요.

shinsanghooon commented on src/test/java/com/shyb/evcharging/user/application/UserServiceTest.java:
```diff
@@ -0,0 +1,211 @@
+        String UPDATE_NAME = "후후후";
+        String UPDATE_PHONE = "010-1234-5678";
+
+        @DisplayName("사용자가 있으면 사용자 정보를 수정한 후 리턴한다.")
```
테스트 코드 작성 질문이 있습니다. 
사용자 정보 수정을 할 때 mybatis를 사용하면 update를 실행할 때 정보가 수정된 사용자 객체를 받아 오는 방법이 없더라구요.(또는 제가 못찾은 걸 수도 있습니다...ㅎㅎ)

그래서 `userService.update()` 메소드 내부에 `userService.update()`에 사용자 정보를 업데이트 하기 전에 `findById()`를 호출해서 등록된 사용자인지 확인하는 코드와 정보 업데이트 이후에 다시 `findById()`를 사용해 업데이트 된 사용자 정보를 조회해서 리턴하는 코드를 작성하였습니다.

이런 상황에서 테스트 코드를 작성하는데 어려움을 겪었습니다. 
등록된 사용자가 존재한다는 것을 알려주기 위해 mock 객체에 대한 조건을 아래처럼 미리 작성해두었는데요. 
`given(userRepository.findById(USER_ID)).willReturn(Optional.of(userResponseDto));`
그런데 이로 인해 userService.update()가 리턴하는 값 또한 userResponseDto가 되는 상황이 발생했습니다. 따라서 사용자 정보가 잘 수정되었는지에 대한 테스트 코드를 작성하지 못하였습니다.

이것저것 찾아보기도 하고 고민도 해보았지만 어떻게 해결하면 좋을지 방법이 떠오르지 않았습니다. 이런 경우 테스트 코드를 어떻게 작성하는게 좋을지 궁금합니다.

shinsanghooon commented on src/test/java/com/shyb/evcharging/user/application/UserServiceTest.java:
```diff
@@ -0,0 +1,211 @@
+        String UPDATE_NAME = "후후후";
+        String UPDATE_PHONE = "010-1234-5678";
+
+        @DisplayName("사용자가 있으면 사용자 정보를 수정한 후 리턴한다.")
+        @Test
+        void update_with_valid() {
+//            // given
+//            UserModifyRequestDto userModifyRequestDto = new UserModifyRequestDto();
+//            userModifyRequestDto.setName(UPDATE_NAME);
+//            userModifyRequestDto.setPhone(UPDATE_PHONE);
+//
+//            given(userRepository.findById(USER_ID))
+//                .willReturn(Optional.of(userResponseDto));
+//
+//            // when
+//            UserResponseDto updatedUser = userService.update(USER_ID, userModifyRequestDto);
+//
+//            // then
+//            assertThat(updatedUser.getName()).isEqualTo(UPDATE_NAME);
+//            assertThat(updatedUser.getPhone()).isEqualTo(UPDATE_PHONE);
+        }
+
```
테스트 코드 작성 질문이 있습니다. 
사용자 정보 수정을 할 때 mybatis를 사용하면 update를 실행할 때 정보가 수정된 사용자 객체를 받아 오는 방법이 없더라구요.(또는 제가 못찾은 걸 수도 있습니다...ㅎㅎ)

그래서 `userService.update()` 메소드 내부에 `userService.update()`에 사용자 정보를 업데이트 하기 전에 `findById()`를 호출해서 등록된 사용자인지 확인하는 코드와 정보 업데이트 이후에 다시 `findById()`를 사용해 업데이트 된 사용자 정보를 조회해서 리턴하는 코드를 작성하였습니다.

이런 상황에서 테스트 코드를 작성하는데 어려움을 겪었습니다. 
등록된 사용자가 존재한다는 것을 알려주기 위해 mock 객체에 대한 조건을 아래처럼 미리 작성해두었는데요. 
`given(userRepository.findById(USER_ID)).willReturn(Optional.of(userResponseDto));`
그런데 이로 인해 userService.update()가 리턴하는 값 또한 userResponseDto가 되는 상황이 발생했습니다. 따라서 사용자 정보가 잘 수정되었는지에 대한 테스트 코드를 작성하지 못하였습니다.

이것저것 찾아보기도 하고 고민도 해보았지만 어떻게 해결하면 좋을지 방법이 떠오르지 않았습니다. 이런 경우 테스트 코드를 어떻게 작성하는게 좋을지 궁금합니다.

kyucumber replied:
> 테스트 코드 작성 질문이 있습니다.
> 사용자 정보 수정을 할 때 mybatis를 사용하면 update를 실행할 때 정보가 수정된 사용자 객체를 받아 오는 방법이 없더라구요.(또는 > 제가 못찾은 걸 수도 있습니다...ㅎㅎ)

> 그래서 userService.update() 메소드 내부에 userService.update()에 사용자 정보를 업데이트 하기 전에 findById()를 호출해서 > 등록된 사용자인지 확인하는 코드와 정보 업데이트 이후에 다시 findById()를 사용해 업데이트 된 사용자 정보를 조회해서 리턴하는 코> 드를 작성하였습니다.

위 케이스는 저도 mybatis와 관련되어서 조금 찾아봐야 할 것 같아서 관련 내용 파악한 뒤 이후에 답변 드릴게요.

> 이런 상황에서 테스트 코드를 작성하는데 어려움을 겪었습니다.

이 케이스의 경우 findById가 여러번 호출되며 이에 따라 테스트가 어려웠다는 말씀인 것 같아요.

```java
given(userRepository.findById(USER_ID))
    .willReturn(Optional.of(userResponseDto))
    .willReturn(Optional.of(updatedUserResponseDto))
```

이런 경우 위처럼 첫번째 호출에는 userResponseDto, 두번째 호출에는 updatedUserResponseDto가 반환되도록 할 수 있었던 것 같아요.

mockito consecutive calls 과 같은 키워드로 한번 찾아보시면 관련해서 나오지 않을까 싶습니다.

shinsanghooon replied:
와우 willReturn 2번 사용하니까 해결이 되었습니다! 알려주신 키워드도 찾아보도록 하겠습니다 감사합니다! 😀

shinsanghooon committed:
refactor: 컨트롤러 path 수정

1. 인증이 적용될 경로와 적용되지 않을 경로를 구분할 필요가 있는 경우를
   대비해 접두사를 추가함 (`/users` -> ``/api/v1/users`)

Related to #8, #9

shinsanghooon committed:
refactor: 이메일 중복 체크 API를 @GetMapping으로 수정

1. 기존 @PostMapping을 @GetMapping으로 변경. 중복 여부를 하나의 자원의
   조회로 바라 볼 수 있기 때문에 Get이 조금 더 의미적으로 알맞는 것 같아
   변경함
2. 이메일 중복 체크 관련 테스트 코드 수정

Related to #8, #9

shinsanghooon committed:
refactor: ErrorResponse를 ErrorResponseDto로 변경

1. ErrorResponse도 에러에 대한 데이터 전달하는 역할을 하는 DTO
   객체이므로 통일성을 위해 ErrorResponseDto로 이름을 변경

shinsanghooon committed:
refactor: ErrorResponse 객체를 ErrorResponseDto 객체로 변경

1. ErrorResponse도 에러에 대한 데이터 전달하는 역할을 하는 DTO 객체이므로 통일성을 위해 ErrorResponseDto로 이름을 변경

shinsanghooon committed:
refactor: UserRequestDTO 불변 수정

1. 패스워드 암호화를 UserRequestDTO 객체의 메소드를 통해 진행함. 내부
   패스워드 필드 값을 변경하기 때문에 불변을 유지하지 못하는 상황이 발생
2. 암호화된 패스워드를 따라 생성한 뒤 생성되는 User 객체에 암호화된
   패스워드를 전달해줌으로 불변을 유지할 수 있게 함

Related to #8

shinsanghooon committed:
fix: 스프링 validation 시 여러 필드에서 예외가 나는 경우 처리

1. 기존 코드에서는 여러 필드에서 예외가 동시에 던져지더라도 하나의
   필드에 대한 예외 메시지만 보여지게 구현되어 있었음
2. 여러 필드 동시 예외 상황을 대비해서 전체 메시지를 ErrorResponseDto에
   담아서 전달할 수 있도록 함

shinsanghooon commented on src/main/java/com/shyb/evcharging/user/application/UserService.java:
```diff
@@ -0,0 +1,94 @@
+     * @param userModifyRequestDto 사용자 정보 요청 DTO
+     * @return 정보가 변경된 사용자
+     */
+    public UserResponseDto update(Long id, UserModifyRequestDto userModifyRequestDto) {
+
+        if (userModifyRequestDto.getPassword() != null && userModifyRequestDto.getConfirmPassword() != null) {
+            boolean sameString = CustomStringUtils.isSameString(userModifyRequestDto.getPassword(), userModifyRequestDto.getConfirmPassword());
+            if (!sameString) {
+                throw new PasswordMisMatchException();
+            }
+            userModifyRequestDto.setPassword(PasswordEncoder.encrypt(userModifyRequestDto.getPassword()));
+        }
+
+        if (findById(id) != null) {
+            userRepository.update(id, userModifyRequestDto);
+        }
+
+        return findById(id);
+    }
```
코멘트 남겨주신 것 중에 DTO 불변에 대한 것이 있었는데, 여기서도 userModifyRequestDto의 password 필드에 암호화된 패스워드 값을 변경해주고 있습니다.

그럼 여기도 다른 리뷰에서 알려주신 방법처럼 User 객체를 만들어서 userRepository.update() 메소드에 전달해주는 좋은 방법인가요? 

그렇게 하면 다른 리뷰에서도 남겨주신 DTO가 Repository를 침범하는 경우를 막을 수 있을 것 같습니다. DTO 객체들은 Repository까지 전달해서 주고 받다보니, User 객체의 필요성에 의문을 가졌었는데 좋지 않은 형태였네요.

kyucumber replied:
네 Dto를 넘기는 건 지양하고 Dto의 데이터를 받아서 user.update()와 같이 User 도메인 클래스 내에 update 등의 메소드를 추가하고 userRepository의 update 메소드에서 업데이트된 User 클래스를 전달받는 형태로 구현하는게 좋을 것 같아요.

shinsanghooon committed:
feat [#9]: 회원 정보 수정 테스트 코드 작성

1. 회원 정보 수정 테스트 코드 작성

Close #9

shinsanghooon committed:
refactor: Repository 레이어에서 DTO를 반환하지 않게 수정

1. Controller 레이어에서 데이터를 주고 받을 목적으로 사용하는 DTO를 Repository 레이어에서 사용하는 것은 의미적으로 맞지 않다고 생각하여 수정
2. Controller, Service, Repository 전반에 걸쳐 DTO 객체가 사용되고
   있으므로 DTO 변경의 경우 모든 레이어가 영향을 받을 수 있게 되는
   경우를 방지할 수 있음

shinsanghooon committed:
refactor: 사용하지 않는 코드 삭제

1. 사용하지 않는 코드 삭제

shinsanghooon committed:
feat [#9]: 사용자 정보 수정 테스트 코드 추가 작성

1. 비밀번호 변경 시 패스워드 일치 및 불일치 상황에 대한 테스트 코드 추가
2. 비밀번호도 변경하는 경우에 대한 테스트 코드 추가

shinsanghooon committed:
refactor: UserModifyRequestDto를 User 객체로 변환해 repository에
전달하도록 수정

1. UserModifyRequestDto를 Repository에 전달하여 쿼리를 업데이트 하게
   구현했으나 Service 레이어에서 DTO 객체의 필드 값을 변경하여 불변을
   지키지 못하게 됨
2. 따라서 repository로 넘겨주기 전에 User 객체로 변환하여 전달하도록
   수정

shinsanghooon committed:
refacotr: 테스트 코드에 Test Fixture 반영

1. 반복되는 Fixture와 더불어 가독성을 위해 TestFixture 클래스에서
   Fixture를 따로 관리할 수 있도록 변경함

f-lab-kyucumber commented:
코멘트 내용 대부분 반영되어서 Approve 했습니다. 확인하고 머지하신 뒤 이어서 진행해주세요~
